### PR TITLE
Preprocessing and DataModule

### PR DESCRIPTION
**Pull request by white-alistair**
This is a work in progress.

I've created a new file `era5_daily.py` that will replace `nc2np_equally_era5_daily.py`. Diffs in other files are just from auto-formatting using black and can be ignored (maybe we should agree on some common settings here?).

One year of daily ERA5 data is approx 250MB so we can just load everything into memory.

### TODO
- [x] Save to disk
- [x] Try to do the merge in-place
- [ ] Decide on correct shape of processed data
- [ ] Normalisation

### CODE REVIEW

**Comment by maximilian-gelbrecht** [comment]
Looks good. Thanks for the fixes. 

Yeah, I agree we can just write simpler data loader than those present in the repo. 

Are you doing that or should I go ahead?

With regards to code formatting: I am fine with whatever setting. Just let me know.

**Comment by white-alistair** [comment]
Working on it :)

**Comment by white-alistair** [comment]
@maximilian-gelbrecht still a few to-do's but I think it's a good time for a review.

File: src/climax/enso/datamodule.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        self.end_year = end_year
+        self.forecast_lead_time = forecast_lead_time
+
+        # Dataloader args
+        self.batch_size = batch_size
+        self.num_workers = num_workers
+        self.pin_memory = pin_memory
+
+        # Set up datasets
+        # TODO: is here the best place to do this?
```
Comment (maximilian-gelbrecht) [inline src/climax/enso/datamodule.py:39]: I didn't use `LightningDataModule`s before but given what's written in the manual, this could also go into the `setup` I think.

File: src/climax/enso/datamodule.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+class ENSODataModule(LightningDataModule):
+    def __init__(
+        self,
+        data_dir: str,
+        start_train_year: int,
+        start_val_year: int,
+        start_test_year: int,
+        end_year: int,
+        forecast_lead_time: int,
```
Comment (maximilian-gelbrecht) [inline src/climax/enso/datamodule.py:18]: `forecast_lead_time` is in days, right?

File: src/climax/enso/datamodule.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            self.start_test_year,
+            self.end_year,
+        )
+
+    def setup(self, stage):
+        pass
+
+    def get_dataset(self, era5_dataset, enso_dataset, start_year, end_year):
+        start_date = date(start_year, 1, 1)
+        end_date = date(end_year, 12, 31)
```
Comment (maximilian-gelbrecht) [inline src/climax/enso/datamodule.py:67]: You removed the leap year treatment, right? I also think we can deal with leap years. Applying the transformers blocks to the time series with one element more should work. What do you think about the time embedding in this case? The climate projection model has a learned time embedding (the lead time embedding is separate). I am not quite sure how to deal with the leap year there, but I am also not sure how important it is.

**Review by maximilian-gelbrecht — COMMENTED** [review]
Looks good! 

Let's try it to train it with this data first, but then we could also consider adding SLP to the data as well. Maybe we have to retrain the embedding anyway with the daily data, then adding further data also wouldn't be an issue.

File: src/climax/enso/datamodule.py (hunk 1 above)
Comment (white-alistair) [reply src/climax/enso/datamodule.py:39]: Yeah I actually had it there but moved it for some reason (as you can see by the empty method). Will move it back.

File: src/climax/enso/datamodule.py (hunk 3 above)
Comment (white-alistair) [reply src/climax/enso/datamodule.py:67]: Since we are dealing with daily data, I don't actually understand why we need to give the leap days special treatment?

**Comment by white-alistair** [comment]
> Looks good!
> 
> Let's try it to train it with this data first, but then we could also consider adding SLP to the data as well. Maybe we have to retrain the embedding anyway with the daily data, then adding further data also wouldn't be an issue.

SLP = sea level pressure?

**Comment by maximilian-gelbrecht** [comment]
> > Looks good!
> > Let's try it to train it with this data first, but then we could also consider adding SLP to the data as well. Maybe we have to retrain the embedding anyway with the daily data, then adding further data also wouldn't be an issue.
> 
> SLP = sea level pressure?

Ups, yes SLP is sea level pressure, which we could also add (it's included in some ENSO indices like MEI). But I was actually thinking about SST sea surface temperature which we also use to compute the index.

File: src/climax/enso/datamodule.py (hunk 3 above)
Comment (maximilian-gelbrecht) [reply src/climax/enso/datamodule.py:67]: Not really special treatment, but it's just a bit unclear how best to do the embedding with it. The climate projection architecture that we could use has a learned embedding for each time step. https://github.com/maximilian-gelbrecht/ClimaX/blob/c039875258a493156c38c878cc62472139f47ed2/src/climax/climate_projection/arch.py#L57

With some inputs being one element longer than others, this just becomes a bit unclear how to do it best. But I think we can also think about some way how to deal with it.

Comment (white-alistair) [reply src/climax/enso/datamodule.py:67]: The bit I don't understand is why some inputs would be one element longer than others; would our input not simply be X days of ERA5 data, where X is some number that we choose?

Writing this also makes me realise that I need to change the dataloader, since it currently just returns one days of ERA5 data at the given lead time.

Comment (maximilian-gelbrecht) [reply src/climax/enso/datamodule.py:67]: Ah yes, sorry, I was somehow thinking of always giving one year (somehow including leap years). But you are totally right, just X days is absolutely the way to go.

Comment (maximilian-gelbrecht) [reply src/climax/enso/datamodule.py:67]: Makes we wonder why they even treated leap years differently in the paper in the first place

Comment (maximilian-gelbrecht) [reply src/climax/enso/datamodule.py:67]: Thinking a bit more about this, in ENSO prediction, there are certain times in the year when it's harder to forecast then in others. So the seasonal dynamics play a role. If I recall correctly, e.g. there's a "spring barrier" through which predictability goes down. So we could think about the temporal embedding still encoding the calendar day in some way. Maybe. But yeah, I thinks that's just one of the many things to play around with.
